File: src/CalcViewModel/Common/TraceLogger.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "TraceLogger.h"
#include "NetworkManager.h"
#include "CalculatorButtonUser.h"

using namespace CalculatorApp;
using namespace CalculatorApp::ViewModel::Common;
using namespace CalculatorApp::ViewModel;
using namespace TraceLogging;
using namespace Concurrency;
using namespace std;
using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Diagnostics;
using namespace Windows::Globalization;
using namespace Windows::Globalization::DateTimeFormatting;
using namespace Windows::System::UserProfile;

namespace CalculatorApp
{
    static constexpr array<const wchar_t* const, 9> s_programmerType{ L"N/A",     L"QwordType", L"DwordType", L"WordType", L"ByteType",
                                                                      L"HexBase", L"DecBase",   L"OctBase",   L"BinBase" };
    static reader_writer_lock s_traceLoggerLock;

    // Diagnostics events. Uploaded to asimov.
    constexpr auto EVENT_NAME_WINDOW_ON_CREATED = L"WindowCreated";
    constexpr auto EVENT_NAME_BUTTON_USAGE = L"ButtonUsageInSession";
    constexpr auto EVENT_NAME_NAV_BAR_OPENED = L"NavigationViewOpened";
    constexpr auto EVENT_NAME_MODE_CHANGED = L"ModeChanged";
    constexpr auto EVENT_NAME_DATE_CALCULATION_MODE_USED = L"DateCalculationModeUsed";
    constexpr auto EVENT_NAME_HISTORY_ITEM_LOAD = L"HistoryItemLoad";
    constexpr auto EVENT_NAME_MEMORY_ITEM_LOAD = L"MemoryItemLoad";
    constexpr auto EVENT_NAME_VISUAL_STATE_CHANGED = L"VisualStateChanged";
    constexpr auto EVENT_NAME_CONVERTER_INPUT_RECEIVED = L"ConverterInputReceived";
    constexpr auto EVENT_NAME_INPUT_PASTED = L"InputPasted";
    constexpr auto EVENT_NAME_SHOW_HIDE_BUTTON_CLICKED = L"ShowHideButtonClicked";
    constexpr auto EVENT_NAME_GRAPH_BUTTON_CLICKED = L"GraphButtonClicked";
    constexpr auto EVENT_NAME_GRAPH_LINE_STYLE_CHANGED = L"GraphLineStyleChanged";
    constexpr auto EVENT_NAME_VARIABLE_CHANGED = L"VariableChanged";
    constexpr auto EVENT_NAME_VARIABLE_SETTING_CHANGED = L"VariableSettingChanged";
    constexpr auto EVENT_NAME_GRAPH_SETTINGS_CHANGED = L"GraphSettingsChanged";
    constexpr auto EVENT_NAME_GRAPH_THEME = L"GraphTheme";
    constexpr auto EVENT_NAME_RECALL_SNAPSHOT = L"RecallSnapshot";
    constexpr auto EVENT_NAME_RECALL_RESTORE = L"RecallRestore";

    constexpr auto EVENT_NAME_EXCEPTION = L"Exception";

    constexpr auto CALC_MODE = L"CalcMode";
    constexpr auto GRAPHING_MODE = L"Graphing";

#pragma region TraceLogger setup and cleanup

    TraceLogger::TraceLogger()
    {
    }

    TraceLogger ^ TraceLogger::GetInstance()
    {
        static TraceLogger ^ s_selfInstance = ref new TraceLogger();
        return s_selfInstance;
    }

    // return true if windowId is logged once else return false
    bool TraceLogger::IsWindowIdInLog(int windowId)
    {
        // Writer lock for the windowIdLog resource
        reader_writer_lock::scoped_lock lock(s_traceLoggerLock);

        if (find(windowIdLog.begin(), windowIdLog.end(), windowId) == windowIdLog.end())
        {
            return false;
        }

        return true;
    }

    void TraceLogger::LogVisualStateChanged(ViewMode mode, String ^ state, bool isAlwaysOnTop)
    {
        auto fields = ref new LoggingFields();

        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        fields->AddString(StringReference(L"VisualState"), state);
        fields->AddBoolean(StringReference(L"IsAlwaysOnTop"), isAlwaysOnTop);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_VISUAL_STATE_CHANGED), fields);
    }

    void TraceLogger::LogWindowCreated(ViewMode mode, int windowId)
    {
        // store windowId in windowIdLog which says we have logged mode for the present windowId.
        if (!IsWindowIdInLog(windowId))
        {
            windowIdLog.push_back(windowId);
        }

        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        fields->AddUInt64(StringReference(L"NumOfOpenWindows"), currentWindowCount);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_WINDOW_ON_CREATED), fields);
    }

    void TraceLogger::LogModeChange(ViewMode mode)
    {
        if (NavCategoryStates::IsValidViewMode(mode))
        {
            auto fields = ref new LoggingFields();
            fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
            TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_MODE_CHANGED), fields);
        }
    }

    void TraceLogger::LogHistoryItemLoad(ViewMode mode, int historyListSize, int loadedIndex)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        fields->AddInt32(StringReference(L"HistoryListSize"), historyListSize);
        fields->AddInt32(StringReference(L"HistoryItemIndex"), loadedIndex);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_HISTORY_ITEM_LOAD), fields);
    }

    void TraceLogger::LogMemoryItemLoad(ViewMode mode, int memoryListSize, int loadedIndex)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        fields->AddInt32(StringReference(L"MemoryListSize"), memoryListSize);
        fields->AddInt32(StringReference(L"MemoryItemIndex"), loadedIndex);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_MEMORY_ITEM_LOAD), fields);
    }

    void TraceLogger::LogError(ViewMode mode, Platform::String ^ functionName, Platform::String ^ errorString)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        fields->AddString(StringReference(L"FunctionName"), functionName);
        fields->AddString(StringReference(L"Message"), errorString);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_EXCEPTION), fields);
    }

    void TraceLogger::LogStandardException(ViewMode mode, wstring_view functionName, const exception& e)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        fields->AddString(StringReference(L"FunctionName"), StringReference(functionName.data()));
        wstringstream exceptionMessage;
        exceptionMessage << e.what();
        fields->AddString(StringReference(L"Message"), StringReference(exceptionMessage.str().data()));
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_EXCEPTION), fields);
    }

    void TraceLogger::LogPlatformExceptionInfo(CalculatorApp::ViewModel::Common::ViewMode mode, Platform::String ^ functionName, Platform::String^ message, int hresult)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        fields->AddString(StringReference(L"FunctionName"), functionName);
        fields->AddString(StringReference(L"Message"), message);
        fields->AddInt32(StringReference(L"HRESULT"), hresult);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_EXCEPTION), fields);
    }

    void TraceLogger::LogPlatformException(ViewMode mode, Platform::String ^ functionName, Platform::Exception ^ e)
    {
        LogPlatformExceptionInfo(mode, functionName, e->Message, e->HResult);
    }

    void TraceLogger::UpdateButtonUsage(NumbersAndOperatorsEnum button, ViewMode mode)
    {
        // IsProgrammerMode, IsScientificMode, IsStandardMode and None are not actual buttons, so ignore them
        if (button == NumbersAndOperatorsEnum::IsProgrammerMode || button == NumbersAndOperatorsEnum::IsScientificMode
            || button == NumbersAndOperatorsEnum::IsStandardMode || button == NumbersAndOperatorsEnum::None)
        {
            return;
        }

        {
            // Writer lock for the buttonLog resource
            reader_writer_lock::scoped_lock lock(s_traceLoggerLock);

            vector<ButtonLog>::iterator it = std::find_if(
                buttonLog.begin(), buttonLog.end(), [button, mode](const ButtonLog& bLog) -> bool { return bLog.button == button && bLog.mode == mode; });
            if (it != buttonLog.end())
            {
                it->count++;
            }
            else
            {
                buttonLog.push_back(ButtonLog(button, mode));
            }
        }

        // Periodically log the button usage so that we do not lose all button data if the app is foricibly closed or crashes
        if (buttonLog.size() >= 10)
        {
            LogButtonUsage();
        }
    }

    void TraceLogger::UpdateWindowCount(uint64 windowCount)
    {
        if (windowCount == 0)
        {
            currentWindowCount--;
            return;
        }
        currentWindowCount = windowCount;
    }

    void TraceLogger::DecreaseWindowCount()
    {
        currentWindowCount = 0;
    }

    void TraceLogger::LogButtonUsage()
    {
        // Writer lock for the buttonLog resource
        reader_writer_lock::scoped_lock lock(s_traceLoggerLock);

        if (buttonLog.size() == 0)
        {
            return;
        }

        Platform::String ^ buttonUsageString;
        for (size_t i = 0; i < buttonLog.size(); i++)
        {
            buttonUsageString += NavCategoryStates::GetFriendlyName(buttonLog[i].mode);
            buttonUsageString += "|";
            buttonUsageString += buttonLog[i].button.ToString();
            buttonUsageString += "|";
            buttonUsageString += buttonLog[i].count;
            if (i != buttonLog.size() - 1)
            {
                buttonUsageString += ",";
            }
        }

        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(L"ButtonUsage"), buttonUsageString);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_BUTTON_USAGE), fields);

        buttonLog.clear();
    }

    void TraceLogger::LogDateCalculationModeUsed(bool AddSubtractMode)
    {
        const wchar_t* calculationType = AddSubtractMode ? L"AddSubtractMode" : L"DateDifferenceMode";
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(ViewMode::Date));
        fields->AddString(StringReference(L"CalculationType"), StringReference(calculationType));
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_DATE_CALCULATION_MODE_USED), fields);
    }

    void TraceLogger::LogConverterInputReceived(ViewMode mode)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_CONVERTER_INPUT_RECEIVED), fields);
    }

    void TraceLogger::LogNavBarOpened()
    {
        auto fields = ref new LoggingFields();
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_NAV_BAR_OPENED), fields);
    }

    void TraceLogger::LogInputPasted(ViewMode mode)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_INPUT_PASTED), fields);
    }

    void TraceLogger::LogShowHideButtonClicked(bool isHideButton)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), StringReference(GRAPHING_MODE));
        fields->AddBoolean(StringReference(L"IsHideButton"), isHideButton);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_SHOW_HIDE_BUTTON_CLICKED), fields);
    }

    void TraceLogger::LogGraphButtonClicked(GraphButton buttonName, GraphButtonValue buttonValue)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), StringReference(GRAPHING_MODE));
        fields->AddInt16(StringReference(L"ButtonName"), static_cast<int16>(buttonName));
        fields->AddInt16(StringReference(L"ButtonValue"), static_cast<int16>(buttonValue));
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_GRAPH_BUTTON_CLICKED), fields);
    }

    void TraceLogger::LogGraphLineStyleChanged(LineStyleType style)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), StringReference(GRAPHING_MODE));
        fields->AddInt16(StringReference(L"StyleType"), static_cast<int16>(style));
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_GRAPH_LINE_STYLE_CHANGED), fields);
    }

    void TraceLogger::LogVariableChanged(String ^ inputChangedType, String ^ variableName)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), StringReference(GRAPHING_MODE));
        fields->AddString(StringReference(L"InputChangedType"), inputChangedType);
        fields->AddString(StringReference(L"VariableName"), variableName);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_VARIABLE_CHANGED), fields);
    }
    void TraceLogger::LogVariableSettingsChanged(String ^ setting)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), StringReference(GRAPHING_MODE));
        fields->AddString(StringReference(L"SettingChanged"), setting);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_VARIABLE_SETTING_CHANGED), fields);
    }

    void TraceLogger::LogGraphSettingsChanged(GraphSettingsType settingType, String ^ settingValue)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), StringReference(GRAPHING_MODE));
        fields->AddInt16(L"SettingType", static_cast<int16>(settingType));
        fields->AddString(L"SettingValue", settingValue);

        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_GRAPH_SETTINGS_CHANGED), fields);
    }

    void TraceLogger::LogGraphTheme(String ^ graphTheme)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), StringReference(GRAPHING_MODE));
        fields->AddString(L"GraphTheme", graphTheme);

        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_GRAPH_THEME), fields);
    }

    void TraceLogger::LogRecallSnapshot(ViewMode mode)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_RECALL_SNAPSHOT), fields);
    }

    void TraceLogger::LogRecallRestore(ViewMode mode)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(CALC_MODE), NavCategoryStates::GetFriendlyName(mode));
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_RECALL_RESTORE), fields);
    }

    void TraceLogger::LogRecallError(Platform::String^ message)
    {
        auto fields = ref new LoggingFields();
        fields->AddString(StringReference(L"FunctionName"), L"Recall");
        fields->AddString(StringReference(L"Message"), message);
        TraceLoggingCommon::GetInstance()->LogLevel2Event(StringReference(EVENT_NAME_EXCEPTION), fields);
    }
}

File: src/Calculator.ViewModels/Common/TraceLogger.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// A trace logging provider can only be instantiated and registered once per module.
// This class implements a singleton model to ensure that only one instance is created.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Windows.Foundation.Diagnostics;

namespace CalculatorApp.ViewModel.Common
{
    public enum GraphSettingsType
    {
        Grid,
        TrigUnits,
        Theme
    }

    public enum GraphButton
    {
        StylePicker,
        RemoveFunction,
        ActiveTracingChecked,
        ActiveTracingUnchecked,
        GraphSettings,
        Share,
        ZoomIn,
        ZoomOut,
        GraphView
    }

    public enum GraphButtonValue
    {
        None,
        AutomaticBestFit,
        ManualAdjustment
    }

    public enum LineStyleType
    {
        Color,
        Pattern
    }

    internal struct ButtonLog
    {
        public int Count;
        public NumbersAndOperatorsEnum Button;
        public ViewMode Mode;

        public ButtonLog(NumbersAndOperatorsEnum button, ViewMode mode)
        {
            Button = button;
            Mode = mode;
            Count = 1;
        }
    }

    public sealed class TraceLogger
    {
        private static readonly Lazy<TraceLogger> s_instance = new Lazy<TraceLogger>(() => new TraceLogger());

        private const string CalcMode = "CalcMode";
        private const string GraphingMode = "Graphing";

        private const string EventNameWindowOnCreated = "WindowCreated";
        private const string EventNameButtonUsage = "ButtonUsageInSession";
        private const string EventNameNavBarOpened = "NavigationViewOpened";
        private const string EventNameModeChanged = "ModeChanged";
        private const string EventNameDateCalculationModeUsed = "DateCalculationModeUsed";
        private const string EventNameHistoryItemLoad = "HistoryItemLoad";
        private const string EventNameMemoryItemLoad = "MemoryItemLoad";
        private const string EventNameVisualStateChanged = "VisualStateChanged";
        private const string EventNameConverterInputReceived = "ConverterInputReceived";
        private const string EventNameInputPasted = "InputPasted";
        private const string EventNameShowHideButtonClicked = "ShowHideButtonClicked";
        private const string EventNameGraphButtonClicked = "GraphButtonClicked";
        private const string EventNameGraphLineStyleChanged = "GraphLineStyleChanged";
        private const string EventNameVariableChanged = "VariableChanged";
        private const string EventNameVariableSettingChanged = "VariableSettingChanged";
        private const string EventNameGraphSettingsChanged = "GraphSettingsChanged";
        private const string EventNameGraphTheme = "GraphTheme";
        private const string EventNameRecallSnapshot = "RecallSnapshot";
        private const string EventNameRecallRestore = "RecallRestore";
        private const string EventNameException = "Exception";

        private readonly List<ButtonLog> _buttonLog = new List<ButtonLog>();
        private readonly List<int> _windowIdLog = new List<int>();
        private readonly object _lock = new object();
        private ulong _currentWindowCount;

        private TraceLogger()
        {
        }

        public static TraceLogger GetInstance() => s_instance.Value;

        public bool IsWindowIdInLog(int windowId)
        {
            lock (_lock)
            {
                return _windowIdLog.Contains(windowId);
            }
        }

        public void LogVisualStateChanged(ViewMode mode, string state, bool isAlwaysOnTop)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            fields.AddString("VisualState", state);
            fields.AddBoolean("IsAlwaysOnTop", isAlwaysOnTop);
            LogLevel2Event(EventNameVisualStateChanged, fields);
        }

        public void LogWindowCreated(ViewMode mode, int windowId)
        {
            if (!IsWindowIdInLog(windowId))
            {
                lock (_lock)
                {
                    _windowIdLog.Add(windowId);
                }
            }

            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            fields.AddUInt64("NumOfOpenWindows", _currentWindowCount);
            LogLevel2Event(EventNameWindowOnCreated, fields);
        }

        public void LogModeChange(ViewMode mode)
        {
            if (NavCategoryStates.IsValidViewMode(mode))
            {
                var fields = new LoggingFields();
                fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
                LogLevel2Event(EventNameModeChanged, fields);
            }
        }

        public void LogHistoryItemLoad(ViewMode mode, int historyListSize, int loadedIndex)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            fields.AddInt32("HistoryListSize", historyListSize);
            fields.AddInt32("HistoryItemIndex", loadedIndex);
            LogLevel2Event(EventNameHistoryItemLoad, fields);
        }

        public void LogMemoryItemLoad(ViewMode mode, int memoryListSize, int loadedIndex)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            fields.AddInt32("MemoryListSize", memoryListSize);
            fields.AddInt32("MemoryItemIndex", loadedIndex);
            LogLevel2Event(EventNameMemoryItemLoad, fields);
        }

        public void LogError(ViewMode mode, string functionName, string errorString)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            fields.AddString("FunctionName", functionName);
            fields.AddString("Message", errorString);
            LogLevel2Event(EventNameException, fields);
        }

        public void LogPlatformExceptionInfo(ViewMode mode, string functionName, string message, int hresult)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            fields.AddString("FunctionName", functionName);
            fields.AddString("Message", message);
            fields.AddInt32("HRESULT", hresult);
            LogLevel2Event(EventNameException, fields);
        }

        internal void LogPlatformException(ViewMode mode, string functionName, Exception e)
        {
            LogPlatformExceptionInfo(mode, functionName, e.Message, e.HResult);
        }

        internal void LogStandardException(ViewMode mode, string functionName, Exception e)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            fields.AddString("FunctionName", functionName);
            fields.AddString("Message", e.Message);
            LogLevel2Event(EventNameException, fields);
        }

        public void UpdateButtonUsage(NumbersAndOperatorsEnum button, ViewMode mode)
        {
            // IsProgrammerMode, IsScientificMode, IsStandardMode and None are not actual buttons, so ignore them
            if (button == NumbersAndOperatorsEnum.IsProgrammerMode
                || button == NumbersAndOperatorsEnum.IsScientificMode
                || button == NumbersAndOperatorsEnum.IsStandardMode
                || button == NumbersAndOperatorsEnum.None)
            {
                return;
            }

            lock (_lock)
            {
                int existingIndex = _buttonLog.FindIndex(b => b.Button == button && b.Mode == mode);
                if (existingIndex >= 0)
                {
                    var entry = _buttonLog[existingIndex];
                    entry.Count++;
                    _buttonLog[existingIndex] = entry;
                }
                else
                {
                    _buttonLog.Add(new ButtonLog(button, mode));
                }
            }

            // Periodically log the button usage so that we do not lose all button data if the app is forcibly closed or crashes
            if (_buttonLog.Count >= 10)
            {
                LogButtonUsage();
            }
        }

        public void LogButtonUsage()
        {
            lock (_lock)
            {
                if (_buttonLog.Count == 0)
                {
                    return;
                }

                var sb = new StringBuilder();
                for (int i = 0; i < _buttonLog.Count; i++)
                {
                    sb.Append(NavCategoryStates.GetFriendlyName(_buttonLog[i].Mode));
                    sb.Append('|');
                    sb.Append(_buttonLog[i].Button.ToString());
                    sb.Append('|');
                    sb.Append(_buttonLog[i].Count);
                    if (i != _buttonLog.Count - 1)
                    {
                        sb.Append(',');
                    }
                }

                var fields = new LoggingFields();
                fields.AddString("ButtonUsage", sb.ToString());
                LogLevel2Event(EventNameButtonUsage, fields);

                _buttonLog.Clear();
            }
        }

        public void UpdateWindowCount(ulong windowCount)
        {
            if (windowCount == 0)
            {
                _currentWindowCount--;
                return;
            }
            _currentWindowCount = windowCount;
        }

        public void DecreaseWindowCount()
        {
            _currentWindowCount = 0;
        }

        public void LogDateCalculationModeUsed(bool addSubtractMode)
        {
            string calculationType = addSubtractMode ? "AddSubtractMode" : "DateDifferenceMode";
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(ViewMode.Date));
            fields.AddString("CalculationType", calculationType);
            LogLevel2Event(EventNameDateCalculationModeUsed, fields);
        }

        public void LogConverterInputReceived(ViewMode mode)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            LogLevel2Event(EventNameConverterInputReceived, fields);
        }

        public void LogNavBarOpened()
        {
            var fields = new LoggingFields();
            LogLevel2Event(EventNameNavBarOpened, fields);
        }

        public void LogInputPasted(ViewMode mode)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            LogLevel2Event(EventNameInputPasted, fields);
        }

        public void LogShowHideButtonClicked(bool isHideButton)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, GraphingMode);
            fields.AddBoolean("IsHideButton", isHideButton);
            LogLevel2Event(EventNameShowHideButtonClicked, fields);
        }

        public void LogGraphButtonClicked(GraphButton buttonName, GraphButtonValue buttonValue)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, GraphingMode);
            fields.AddInt16("ButtonName", (short)buttonName);
            fields.AddInt16("ButtonValue", (short)buttonValue);
            LogLevel2Event(EventNameGraphButtonClicked, fields);
        }

        public void LogGraphLineStyleChanged(LineStyleType style)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, GraphingMode);
            fields.AddInt16("StyleType", (short)style);
            LogLevel2Event(EventNameGraphLineStyleChanged, fields);
        }

        public void LogVariableChanged(string inputChangedType, string variableName)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, GraphingMode);
            fields.AddString("InputChangedType", inputChangedType);
            fields.AddString("VariableName", variableName);
            LogLevel2Event(EventNameVariableChanged, fields);
        }

        public void LogVariableSettingsChanged(string setting)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, GraphingMode);
            fields.AddString("SettingChanged", setting);
            LogLevel2Event(EventNameVariableSettingChanged, fields);
        }

        public void LogGraphSettingsChanged(GraphSettingsType settingType, string settingValue)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, GraphingMode);
            fields.AddInt16("SettingType", (short)settingType);
            fields.AddString("SettingValue", settingValue);
            LogLevel2Event(EventNameGraphSettingsChanged, fields);
        }

        public void LogGraphTheme(string graphTheme)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, GraphingMode);
            fields.AddString("GraphTheme", graphTheme);
            LogLevel2Event(EventNameGraphTheme, fields);
        }

        public void LogRecallSnapshot(ViewMode mode)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            LogLevel2Event(EventNameRecallSnapshot, fields);
        }

        public void LogRecallRestore(ViewMode mode)
        {
            var fields = new LoggingFields();
            fields.AddString(CalcMode, NavCategoryStates.GetFriendlyName(mode));
            LogLevel2Event(EventNameRecallRestore, fields);
        }

        public void LogRecallError(string message)
        {
            var fields = new LoggingFields();
            fields.AddString("FunctionName", "Recall");
            fields.AddString("Message", message);
            LogLevel2Event(EventNameException, fields);
        }

        private void LogLevel2Event(string eventName, LoggingFields fields)
        {
            // Use TraceLoggingCommon if available, otherwise use LoggingChannel
            try
            {
                TraceLoggingCommon.GetInstance()?.LogLevel2Event(eventName, fields);
            }
            catch
            {
                // Silently ignore trace logging failures
            }
        }
    }

    /// <summary>
    /// Wrapper for TraceLogging infrastructure. In the managed port, this uses the Windows.Foundation.Diagnostics
    /// logging channel.
    /// </summary>
    internal sealed class TraceLoggingCommon
    {
        private static TraceLoggingCommon s_instance;
        private readonly LoggingChannel _channel;

        private TraceLoggingCommon()
        {
            _channel = new LoggingChannel("CalculatorApp", null, new Guid("1d9dda64-be84-57c4-8154-e01299f79f4a"));
        }

        public static TraceLoggingCommon GetInstance()
        {
            if (s_instance == null)
            {
                s_instance = new TraceLoggingCommon();
            }
            return s_instance;
        }

        public void LogLevel2Event(string eventName, LoggingFields fields)
        {
            _channel.LogEvent(eventName, fields, LoggingLevel.Verbose);
        }
    }
}
